### PR TITLE
Support preconfigured_waf_config field on rules for google_compute_security_policy resource

### PR DESCRIPTION
**Pull request by davidalger**
* fixes https://github.com/hashicorp/terraform-provider-google/issues/12735

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `preconfigured_waf_config` block to `google_compute_security_policy` resource (beta)
```

### CODE REVIEW

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 1]
```diff
@@ -395,6 +445,33 @@ func resourceComputeSecurityPolicy() *schema.Resource {
...
+func resourceComputeSecurityPolicyRulePreconfiguredWafConfigExclusionFieldParamsSchema(description string) *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"operator": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					Default:      "EQUALS",
```
Comment (shuyama1) [inline mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:458]: Question: Is this the fixed default value in the API? If so, is there a documentation about this?
Same for the "value" field below.

Comment (davidalger) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:458]: The [API documentation](https://cloud.google.com/compute/docs/reference/rest/beta/securityPolicies) doesn't give much simply specifying that op is an enum and val a string. The gcloud documentation for [add-preconfig-waf-exclusion](https://cloud.google.com/sdk/gcloud/reference/beta/compute/security-policies/rules/add-preconfig-waf-exclusion) includes the following additional detail:

> A field value must be given if the field operator is not EQUALS_ANY, and cannot be given if the field operator is EQUALS_ANY.

So to answer the question: No, an op of `EQUALS` is not a fixed default in the API (and I believe you'll get an error if you don't specify the op), but seemed to be the sensible default since operator must be set when specifying a value and it seems rather idiomatic for value to be an exact match when the operator is not explicitly defined in HCL. If you'd rather change operator to a required field with no default, I'd be ok with that.

The default of `""` for the value is effectively a no-op there to support the requirement that value not be set when specifying an operator of EQUALS_ANY. If value is not specified and operator is anything other than EQUALS_ANY, the API will return the following error:

```
Error updating SecurityPolicy "ptlabs-waf-policy": googleapi: Error 400: Invalid value for field 'resource.preconfiguredWafConfig.exclusions[0].requestQueryParamsToExclude[0].val': ''. The field value must be non-empty if the field operator is not EQUALS_ANY., invalid
```

So value must be an optional field on the block, allowing for example (where the default value of `""` is silently ignored and does nothing):

```hcl
              request_cookie {
                  operator = "EQUALS_ANY"
              }
```

Comment (shuyama1) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:458]: Thanks for the clear explanation! 
In this case, I would recommend that we set `operator` as a required field (it will also prevent users from specifying an empty block which sometimes ends up causing problems in Terraform) and set `value` as an optional field without any default values, and the API side throw the error if users do not specify the `value` field when `operator` is not `EQUALS_ANY`.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 2]
```diff
@@ -150,6 +150,56 @@ func resourceComputeSecurityPolicy() *schema.Resource {
 							Description: `A match condition that incoming traffic is evaluated against. If it evaluates to true, the corresponding action is enforced.`,
 						},
 
+						<% unless version == 'ga' -%>
+						"preconfigured_waf_config": {
```
Comment (shuyama1) [inline mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:154]: We usually want to prevent users from specify empty blocks, so I would recommend that we make the only subfield `exclusion` required. However, it is not a requirement, as I see you set an empty block `preconfigured_waf_config{}` in the test and assume that empty blocks may be needed for some purposes in this case?

Comment (davidalger) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:154]: That is correct, the empty block is required here in order to clear what's inside it from the API object as well as avoid perma-drift. Originally I had `exclusion` marked as required, but changed it to optional so that the provider will cover the edge cases here. And that's why the test is there with the empty block, because I ran into this issue myself and so added the test when setting `exclusion` as optional to ensure it's possible to remove the exclusions from the rule properly.

It's a bit weird, but it's due the `preconfiguredWafConfig` field not existing on the API object until you first set it, but once included the only way to remove the exclusions it contains is to actually send an empty object, after-which `preconfiguredWafConfig: {}` is what will come back in the YAML output from describe.

**Review by shuyama1 — APPROVED** [review]
LGTM! Thanks! Will merge the PR after the tests passed.

File: mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown [hunk 3]
```diff
@@ -111,6 +111,8 @@ The following arguments are supported:
 * `match` - (Required) A match condition that incoming traffic is evaluated against.
     If it evaluates to true, the corresponding `action` is enforced. Structure is [documented below](#nested_match).
 
+* `preconfigured_waf_config` - (Optional) Preconfigured WAF configuration to be applied for the rule. If the rule does not evaluate preconfigured WAF rules, i.e., if evaluatePreconfiguredWaf() is not used, this field will have no effect. Structure is [documented below](#nested_preconfigured_waf_config).
```
Comment (shuyama1) [inline mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown:114]:
```suggestion
* `preconfigured_waf_config` - (Optional, Beta) Preconfigured WAF configuration to be applied for the rule. If the rule does not evaluate preconfigured WAF rules, i.e., if evaluatePreconfiguredWaf() is not used, this field will have no effect. Structure is [documented below](#nested_preconfigured_waf_config).
```
Sorry, just realize that we still to specify this is beta only feature in the doc

File: mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown [hunk 4]
```diff
@@ -147,6 +149,37 @@ The following arguments are supported:
...
+
+  * EQUALS: The operator matches if the field value equals the specified value.
+  * STARTS_WITH: The operator matches if the field value starts with the specified value.
+  * ENDS_WITH: The operator matches if the field value ends with the specified value.
+  * CONTAINS: The operator matches if the field value contains the specified value.
+  * EQUALS_ANY: The operator matches if the field value is any value.
+
+* `value` - (Optional) A request field matching the specified value will be excluded from inspection during preconfigured WAF evaluation.
+    The field value must be given if the field `operator` is not "EQUALS_ANY", and cannot be given if the field `operator` is "EQUALS_ANY".
+
```
Comment (shuyama1) [inline mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown:182]: Also need to add `Beta` label to beta-only fields

**Review by shuyama1 — CHANGES_REQUESTED** [review]
Sorry, still need a small change: we need to specify beta only in the doc

File: mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown (hunk 3 above)
Comment (davidalger) [reply mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown]: Good catch! Beta indicator has been added here.

File: mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown (hunk 4 above)
Comment (davidalger) [reply mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown:182]: The other examples I see of this only add the `Beta` qualifier (with a link) to the outermost block for the beta functionality, so I emulated that adding it only to `preconfigured_waf_config` block which the rest of this is contained by. Lmk if this is sufficient. Thanks!

Comment (shuyama1) [reply mmv1/third_party/terraform/website/docs/r/compute_security_policy.html.markdown:182]: I think it sounds reasonable! Thanks for updating the doc!

**Review by shuyama1 — APPROVED** [review]
